### PR TITLE
Fix for logs not written in the experiment dir

### PR DESCRIPTION
**Pull request by BloodAxe**
This PR changes logging logic a little:
1) On start of the program, it will save logs to `~/sg_logs/last_{pid}.log` file and console.
Default log levels are DEBUG for file and INFO for console. Yet they may be changed via `LOG_LEVEL` and `CONSOLE_LOG_LEVEL` environment variables and also by setting log level explicitly while calling `get_logger`.

2) Once experiment dir is set, BaseSGLogger will redirect these logs to new file under this directory via call `AutoLoggerConfig.setup_logging(filename=self.log_full_file_path, copy_already_logged_messages=True)`

3) Regardless of the state of the program one can always access the current log file (to upload it to Deci) via `AutoLoggerConfig.get_log_file_path().

4) Drop `save_full_train_log` argument as obsolete

### CODE REVIEW

File: src/super_gradients/common/sg_loggers/base_sg_logger.py [hunk 1]
```diff
@@ -102,27 +114,39 @@ def _make_dir(self):
...
+        time_string = time.strftime("%m_%d_%H_%M_%S", time.localtime())
+        # There are two log files, since the regular log_file_path used for `manual` logging of configs/other info
+        # While the log_full_file_path serves as a sink for all SG-related events.
+        # We can combine these two by changing _write_to_log_file implementation to call getLogger().info(...)
+        # if that's would be ok
+        # TODO: Check with Ofri
+        self.log_file_path = f"{self._local_dir}/log_{time_string}.txt"
+        self.log_full_file_path = f"{self._local_dir}/log_full_{time_string}.txt"
+        AutoLoggerConfig.log_everything_to_file(self.log_full_file_path)
+        AutoLoggerConfig.add_console_logging(logging.WARNING)
```
Comment (Louis-Dupont) [inline src/super_gradients/common/sg_loggers/base_sg_logger.py:126]: There are some  INFO logs that I think are nice to see by default.
Why not set it by default to INFO, and add the possibility to set the level with an env variable, like it is done here ? : https://github.com/Deci-AI/super-gradients/blob/0fa438b73bf0a3812910245c2e1ee1aefb50391c/src/super_gradients/common/abstractions/abstract_logger.py#L9

File: src/super_gradients/common/auto_logging/auto_logger.py [hunk 2]
```diff
@@ -1,108 +1,79 @@
...
+            copy_already_logged_messages=False,
+            log_level=log_level,
+        )
+
+    @classmethod
+    def setup_logging(self, filename: str, copy_already_logged_messages: bool, log_level: str = None) -> None:
+        if copy_already_logged_messages and self.filename is not None and os.path.exists(self.filename):
+            with open(self.filename, "r", encoding="utf-8") as src:
+                with open(filename, "w") as dst:
+                    dst.write(src.read())
```
Comment (Louis-Dupont) [inline src/super_gradients/common/auto_logging/auto_logger.py:30]: I think that we should first check if the directory exists with something like:
Path(filename).mkdir(parents=True, exist_ok=True)

OR do it only in "setup_default_logging", because when we call this function from "base_sg_logger" we can assume that "self._local_dir" already exists

Comment (BloodAxe) [reply src/super_gradients/common/auto_logging/auto_logger.py:49]: Agree, thanks for the suggestion.

**Review by ofrimasad — COMMENTED** [review]
few minor comments

File: src/super_gradients/common/abstractions/abstract_logger.py [hunk 3]
```diff
@@ -1,29 +1,18 @@
...
-def get_logger(
-    logger_name: str, training_log_path=None, logs_dir_path=None, log_level=DEFAULT_LOGGING_LEVEL
-) -> logging.Logger:
-    config_dict = AutoLoggerConfig.generate_config_for_module_name(
-        module_name=logger_name, training_log_path=training_log_path, logs_dir_path=logs_dir_path, log_level=log_level
-    )
-    logging.config.dictConfig(config_dict)
+def get_logger(logger_name: str, log_level: Union[str, None] = None) -> logging.Logger:
+    AutoLoggerConfig.getInstance()
     logger: logging.Logger = logging.getLogger(logger_name)
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/common/abstractions/abstract_logger.py:11]: I honestly don't see any problem here. Log files are intended to be verbose and to avoid any ambiguity in them. If we use only the last part of the logger name, we may run into name clash issue. For instance, user has `utils.py` file of his own, in SG we may have `utils.py` file as well, and hydra also has `utils.py`

Now compare imaginary output using full names:
```bash
[2022-11-03 12:44:37] DEBUG - hydra.core.utils - Setting JobRuntime:name=UNKNOWN_NAME
[2022-11-03 12:44:38] WARNING - super_gradients.sanity_check.utils - Environment check unsuccessful - OS not supported
[2022-11-03 12:44:39] ERROR- customer_repo.train.utils - DATASET_PATH not found
```

To short names:
```bash
[2022-11-03 12:44:37] DEBUG - utils - Setting JobRuntime:name=UNKNOWN_NAME
[2022-11-03 12:44:38] WARNING - utils - Environment check unsuccessful - OS not supported
[2022-11-03 12:44:39] ERROR- utils - DATASET_PATH not found
```

To me, the latter option looks more cryptic and untraceable.

What we can do, is change the way we create the loggers in SG.
Currently there are `logger = get_logger(__name__)`. What we can use is `logger = get_logger("SG.Trainer") instead`. But I'm, not convinced we should do this at all as logs are not meant to be neither small nor human-readable. I'm happy to discuss this)

[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/common/abstractions/abstract_logger.py:11]: There are also `%(filename)s` and `%(module)s` variables that we can use btw

File: src/super_gradients/common/auto_logging/auto_logger.py [hunk 4]
```diff
@@ -1,108 +1,113 @@
...
-
-            :return: python dict() with the new auto_logging for the module
+        Setup default logging configuration. Usually happens when app starts, and we don't have
+        experiment dir yet.
+        The default log directory will be `~/sg_logs`
+        :param log_level: The default log level to use. If None, uses LOG_LEVEL and CONSOLE_LOG_LEVEL environment vars.
+        :return: None
         """
+        self._setup_logging(
+            filename=os.path.expanduser(f"~/sg_logs/last_{os.getppid()}.log"),
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/common/auto_logging/auto_logger.py]: I've added the timestamp alongside with ppid.
There is no _easy_ way to log all events to a single file, when using DDP or DataLoader with num_workers > 1 on Windows platform. 
In both these cases a multiple processes will be spawned and multiple logs may be created.
Therefore the log file will have the parent PID to being able to discriminate the logs corresponding to a single run.

To be resolved in future: How we ensure log rotation in `~/sg_logs` to avoid this directory blow up in size over time? (Best idea so far: Find all logs through regex matching and retain only last N files or at most M bytes)

File: src/super_gradients/common/abstractions/abstract_logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/common/abstractions/abstract_logger.py:11]: Example of file logging `%(asctime)s %(levelname)s - %(name)s - %(message)s`
```
[2022-11-07 11:15:15] DEBUG - hydra.core.utils - Setting JobRuntime:name=app
[2022-11-07 11:15:16] DEBUG - super_gradients.training.utils.sg_trainer_utils - "events.out.tfevents.1667812390.Bloodaxe-PC.22104.0" will not be deleted
[2022-11-07 11:15:58] INFO - super_gradients.training.utils.sg_trainer_utils - TRAINING PARAMETERS:
    - Mode:                         Single GPU
    - Number of GPUs:               1          (1 available on the machine)
    - Dataset size:                 100000     (len(train_set))
    - Batch size per GPU:           128        (batch_size)
    - Batch Accumulate:             1          (batch_accumulate)
    - Total batch size:             128        (num_gpus * batch_size)
    - Effective Batch size:         128        (num_gpus * batch_size * batch_accumulate)
    - Iterations per epoch:         781        (len(train_set) / total_batch_size)
    - Gradient updates per epoch:   781        (len(train_set) / effective_batch_size)

[2022-11-07 11:15:58] INFO - super_gradients.training.sg_trainer.sg_trainer - Started training for 400 epochs (0/399)
```

Example of console logging `%(asctime)s %(levelname)s - %(filename)s - %(message)s`
```
[2022-11-07 11:22:34]  WARNING - redirects.py - NOTE: Redirects are currently not supported in Windows or MacOs.
[2022-11-07 11:22:35]  WARNING - callbacks.py - Failed to import deci_lab_client
[2022-11-07 11:22:35]  WARNING - deci_platform_sg_logger.py - Failed to import deci_lab_client
C:\Anaconda3\envs\sg\lib\site-packages\_distutils_hack\__init__.py:33: UserWarning: Setuptools is replacing distutils.
```

[the remark this answers is not in the document]
Comment (ofrimasad) [reply src/super_gradients/common/abstractions/abstract_logger.py:11]: ok. let's use the filename

**Review by ofrimasad — APPROVED** [review]
LGTM :)
